primitives/staking/src/lib.rs

```diff
@@ -69,10 +75,55 @@ pub struct Stake<T: StakingInterface + ?Sized> {
 	/// This is only guaranteed to reflect the amount locked by the staking system. If there are
 	/// non-staking locks on the bonded pair's balance this amount is going to be larger in
 	/// reality.
-	pub total: T::Balance,
+	pub total: Balance,
 	/// The total amount of the stash's balance that will be at stake in any forthcoming
 	/// rounds.
-	pub active: T::Balance,
+	pub active: Balance,
+}
+
+/// A generic staking event listener. Currently used for implementations involved in stake tracking.
+/// Note that the interface is designed in a way that the events are fired post-action, so any
+/// pre-action data that is needed needs to be passed to interface methods.
+/// The rest of the data can be retrieved by using `StakingInterface`.
+pub trait OnStakingUpdate<AccountId, Balance> {
+	/// Track ledger updates.
+	fn on_update_ledger(who: &AccountId, prev_stake: Option<Stake<AccountId, Balance>>);
+	/// Track nominators, those reinstated and also new ones.
+	fn on_nominator_add(who: &AccountId, prev_nominations: Vec<AccountId>);
+	/// Track validators, those reinstated and new.
+	fn on_validator_add(who: &AccountId);
+	/// Track removed validators. Either chilled or those that became nominators instead.
+	fn on_validator_remove(who: &AccountId); // only fire this event when this is an actual Validator
+	/// Track removed nominators.
+	fn on_nominator_remove(who: &AccountId, nominations: Vec<AccountId>); // only fire this if this is an actual Nominator
+	/// Track those participants of staking system that are kicked out for whatever reason.
+	fn on_reaped(who: &AccountId); // -> basically `kill_stash`
```

Comment:
```suggestion
	/// Fired when the stake amount of someone updates.
	///
	/// Also called when someone stakes for the first time. (TODO: is it? this is why we need unit tests for this pallet alone). 
	/// 
	/// This is effectively any changes to the bond amount, such as bonding more funds, and unbonding. 
	fn on_stake_update(who: &AccountId, prev_stake: Option<Stake<AccountId, Balance>>);
	/// Fired when someone sets their intention to nominate, either new, or existing one.
	fn on_nominator_update(who: &AccountId, prev_nominations: Vec<AccountId>);
	/// Fired when someone sets their intention to validate, either new, or existing one. 
	fn on_validator_add(who: &AccountId);
	/// Fired when someone removes their intention to validate, either due to chill or nominating.
	fn on_validator_remove(who: &AccountId); // only fire this event when this is an actual Validator
	/// Fired when someone removes their intention to nominate, either due to chill or validating.
	fn on_nominator_remove(who: &AccountId, nominations: Vec<AccountId>); // only fire this if this is an actual Nominator
	/// fired when someone is fully unstakes.
	fn on_unstake(who: &AccountId); // -> basically `kill_stash`
```

the keyword "ledger" should not be used here as it is an implementation detail of staking.

Reply:
I partially disagree, because we are calling those methods only within staking, and `on_ledger_update` is called only in update_ledger. It makes no sense to have a more confusing name for it then?

Reply:
I would argue that since `EventListener` is not a generic interface, but rather Staking-specific - it's totally fine to have staking-specific names that indeed refer to the implementation. However, I don't believe this deserves the back and forth so I'll just rename them as suggested.